### PR TITLE
Pass ideep:lowp_kind to matmul_forward::compute on cache misses

### PR DESCRIPTION
**Pull request by fadara01**
Optimized dynamic quantization for aarch64 was enabled by #126687 and #134897 

This PR fixes an issue for aarch64 where on a [cache miss](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L592) (e.g. if input dimensions change) [ideep::matmul_forward::compute ](https://github.com/intel/ideep/blob/pytorch-rls-v3.5.3-2/include/ideep/operators/matmul.hpp#L160) (wrongly) runs with the [default lowp_kind (u8s8)](https://github.com/intel/ideep/blob/pytorch-rls-v3.5.3-2/include/ideep/operators/matmul.hpp#L174) which is not supported by oneDNN+ACL (Arm Compute Library), causing the workload to fall back to a much slower oneDNN gemm:jit kernel

Example: 
```python
import torch

DIM = 4096
INPUT_SIZE1 = 32
INPUT_SIZE2 = 16

class LinearNet(torch.nn.Module):
   def __init__(self):
        super().__init__()
        self.fc1 = torch.nn.Linear(DIM, DIM, bias=False)

   def forward(self, x):
        x = self.fc1(x)
        return x

input1 = torch.randn(size=(INPUT_SIZE1, DIM))
input2 = torch.randn(size=(INPUT_SIZE2, DIM))

with torch.no_grad():
    model = LinearNet()
    model =  torch.ao.quantization.quantize_dynamic(model,{torch.nn.Linear})
    
    model(input1)   # this goes to ACL lowp_gemm
    print("="*50)
    model(input2)   # this goes to gemm:jit without this PR, and to ACL with this PR
```
In the code snippet above:
- The matmul from `model(input1)` goes to oneDNN+ACL (in both cases, with and without the PR)
- The matmul from `model(input2)`: **Without this PR**: there's a cache miss (different input shapes) and matmul_forward::compute is run with the default lowp_kind (u8s8). Hence the matmul falls back to gemm:jit in oneDNN. However, **With this PR** the matmul goes to oneDNN+ACL which is around 10x faster than oneDNN+jit.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @milpuz01 @aditew01 @nikhil-arm

### CODE REVIEW

**Comment by fadara01** [comment]
cc @malfet @atalman @jondea  @cfRod @milpuz01  Please kindly review.

**Review by jondea — APPROVED** [review]
Great find, thank you!

**Comment by malfet** [comment]
> @malfet We cant seem to add CI labels

@cfRod you'll need to approve workflow run first..

**Review by malfet — APPROVED** [review]
Thank you for the detailed PR description, my only question will it still work (although slowly) on older ARMv8 platforms like Cortex A75?

**Comment by fadara01** [comment]
> @malfet Thank you for the detailed PR description, my only question will it still work (although slowly) on older ARMv8 platforms like Cortex A75?

Yes, it should work for older Arm platforms too.

**Comment by fadara01** [comment]
@malfet could we please get a reproducer for the regression on x86?
My understanding is that [ideep::forward_matmul::compute](https://github.com/intel/ideep/blob/pytorch-rls-v3.5.3-2/include/ideep/operators/matmul.hpp#L174) will be called with the default arguments on x86 with and without this PR.

**Comment by fadara01** [comment]
> I'm not very familiar with this code, but wasn't it calling https://github.com/intel/ideep/blob/pytorch-rls-v3.5.3/include/ideep/operators/matmul.hpp#L63 before your PR, wasn't it?

@malfet  I don't think this is the case, as the ideep::matmul_forward::compute function you mentioned does not match the arguments we're passing from [qlinear_dynamic.cpp](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L593) in the cache miss (else) branch.

Are you using a vanilla wheel for your internal tests with x86?

If I pip install a wheel on x86, this path (ideep/oneDNN) is not even selected for dynamic quantization (can be deduced by the lack of oneDNN verbose when running with the environment variable ONEDNN_VERBOSE=1). 
I think the fbgemm path is chosen instead.

**Review by jgong5 — APPROVED** [review]

**Comment by fadara01** [comment]
@malfet , Any updates or thoughts?

**Comment by fadara01** [comment]
@malfet, I built torch on x86 with `USE_FBGEMM=0` to start exercising this ideep path and confirmed with print statements that [this function  ](https://github.com/intel/ideep/blob/pytorch-rls-v3.5.3-2/include/ideep/operators/matmul.hpp#L160) gets called before and after my PR.

Could you please have a more serious look at this? It brings 10x speedups for LLMs on aarch64, since all matmuls after prefill are currently getting dispatched to sub-optimal implementations due to the [wrong] default u8s8 lowp_kind.

**Comment by robert-hardwick** [comment]
Raised an issue that we have seen which will be fixed by this change https://github.com/pytorch/pytorch/issues/145216

Not a regression. Those tests have not been enabled before in CI.

**Comment by robert-hardwick** [comment]
i think the rebase failed because this has been previously merged and subsequently reverted
